lib/Runtime/Library/IntlEngineInterfaceExtensionObject.cpp

```diff
@@ -1665,118 +1697,134 @@ namespace Js
     }
 #endif
 
-#ifdef INTL_ICU
-    static int CompareStringICU(_In_opt_count_(langtagLen) const char16 *langtag, _In_ charcount_t langtagLen, _In_z_ const char16 *left, _In_ charcount_t cchLeft, _In_z_ const char16 *right, _In_ charcount_t cchRight,
-        _In_ CollatorSensitivity sensitivity, _In_ bool ignorePunctuation, _In_ bool numeric, _In_ CollatorCaseFirst caseFirst, _Out_ HRESULT *hr)
+    Var IntlEngineInterfaceExtensionObject::EntryIntl_LocaleCompare(RecyclableObject* function, CallInfo callInfo, ...)
     {
-        ASSERT_ENUM(CollatorSensitivity, sensitivity);
-        ASSERT_ENUM(CollatorCaseFirst, caseFirst);
-        Assert(left != nullptr && right != nullptr && hr != nullptr);
-        int ret = 0;
-        *hr = E_INVALIDARG;
-        UErrorCode error = U_ZERO_ERROR;
+#ifdef INTL_WINGLOB
+        AssertOrFailFastMsg(false, "platform.localeCompare should not be called in Intl-WinGlob");
+        return nullptr;
+#else
+        EngineInterfaceObject_CommonFunctionProlog(function, callInfo);
+        INTL_CHECK_ARGS(args.Info.Count == 4 && JavascriptString::Is(args[1]) && JavascriptString::Is(args[2]) && DynamicObject::Is(args[3]));
 
-        char localeID[ULOC_FULLNAME_CAPACITY] = { 0 };
-        int32_t length = 0;
-        if (langtag == nullptr)
+        JavascriptString *left = JavascriptString::UnsafeFromVar(args[1]);
+        JavascriptString *right = JavascriptString::UnsafeFromVar(args[2]);
+        DynamicObject *state = DynamicObject::UnsafeFromVar(args[3]);
+
+        // Below, we lazy-initialize the backing UCollator on the first call to localeCompare
+        // On subsequent calls, the UCollator will be cached in state.hiddenObject
+        // TODO(jahorto): Make these property IDs sane, so that hiddenObject doesn't have different meanings in different contexts
```

Comment:
Agreed, hiddenObject is a super generic term. Can we just call it collator or something?